sae_lens/saes/sae_base.py

```diff
@@ -0,0 +1,588 @@
+            "jumprelu_bandwidth": self.jumprelu_bandwidth,
+        }
+
+    # this needs to exist so we can initialize the parent sae cfg without the training specific
+    # parameters. Maybe there's a cleaner way to do this
+    def get_base_sae_cfg_dict(self) -> dict[str, Any]:
+        return {
+            "architecture": self.architecture,
+            "d_in": self.d_in,
+            "d_sae": self.d_sae,
+            "activation_fn": self.activation_fn,
+            "activation_fn_kwargs": self.activation_fn_kwargs,
+            "apply_b_dec_to_input": self.apply_b_dec_to_input,
+            "dtype": self.dtype,
+            "model_name": self.model_name,
+            "hook_name": self.hook_name,
+            "hook_layer": self.hook_layer,
+            "hook_head_index": self.hook_head_index,
+            "device": self.device,
+            "context_size": self.context_size,
+            "prepend_bos": self.prepend_bos,
+            "finetuning_scaling_factor": self.finetuning_scaling_factor,
+            "normalize_activations": self.normalize_activations,
+            "dataset_path": self.dataset_path,
+            "dataset_trust_remote_code": self.dataset_trust_remote_code,
+            "sae_lens_training_version": self.sae_lens_training_version,
+            "model_from_pretrained_kwargs": self.model_from_pretrained_kwargs,
+            "seqpos_slice": self.seqpos_slice,
+            "neuronpedia_id": self.neuronpedia_id,
+        }
+
+
+class BaseTrainingSAE(BaseSAE, ABC):
+    """Abstract base class for training versions of SAEs."""
+
+    cfg: "TrainingSAEConfig"  # Quote marks make this a forward reference
+
+    def __init__(self, cfg: TrainingSAEConfig, use_error_term: bool = False):
+        super().__init__(cfg, use_error_term)
+
+        # Turn off hook_z reshaping for training mode - the activation store
+        # is expected to handle reshaping before passing data to the SAE
+        self.turn_off_forward_pass_hook_z_reshaping()
+
+        self.mse_loss_fn = self._get_mse_loss_fn()
+
+    @abstractmethod
+    def encode_with_hidden_pre(
+        self, x: Float[torch.Tensor, "... d_in"]
+    ) -> Tuple[Float[torch.Tensor, "... d_sae"], Float[torch.Tensor, "... d_sae"]]:
+        """Encode with access to pre-activation values for training."""
+        pass
+
+    @abstractmethod
+    def calculate_aux_loss(
+        self,
+        feature_acts: torch.Tensor,
+        hidden_pre: torch.Tensor,
+        dead_neuron_mask: Optional[torch.Tensor] = None,
+        current_l1_coefficient: Optional[float] = None,
+        sae_in: Optional[torch.Tensor] = None,
+        sae_out: Optional[torch.Tensor] = None,
+        **kwargs: Any,
+    ) -> torch.Tensor:
+        """Calculate architecture-specific auxiliary loss terms."""
+        pass
+
+    def training_forward_pass(
+        self,
+        sae_in: torch.Tensor,
+        current_l1_coefficient: float,
+        dead_neuron_mask: Optional[torch.Tensor] = None,
```

Comment:
nit: it would be better to pass a param called e.g. `ctx: TrainingContext` that has all the context around whatever current coefficients are for l1 (if it's a standard SAE), l0 (if jumprelu), dead neurons, current training step, whatever else